Refactor Species Class

enjyashraf18 opened the pull request:
To address Issue #148, the handling of atomic species data in AtomDB has been refactored to replace the previous MessagePack-based storage system with a structured HDF5 format.

**Modifications**
1. Atomic datasets are now stored in `datasets_data.h5` file under the /Datasets group, with hierarchical organization. For example:
```bash
/Datasets
├── /slater
│   ├── /H
│   │   └── /H_000_002_000
│   ├── /C
│   │   └── /C_000_003_000
│ 
├── /gaussian
├── /hci
├── /nist
├── /numeric
├── /uhf_augccpvdz
```
2. Enhanced Query Capabilities
- Supports efficient filtering by atomic properties.
- Replaces old searches techniques with direct HDF5 path queries.

3. HDF5 Backend Infrastructure
Created `h5file_creator.py` as the core module for generating the HDF5 structure. It creates organized folders for any atomic species with defined properties in `datasets_data.h5` file, replacing the old MessagePack approach.

4. Class Structure
The `SpeciesData` dataclass has been removed and replaced with a dynamic `DefinitionClass` imported per dataset.

The current modifications focus on the Slater dataset, and all MessagePack files for this dataset have been successfully migrated to `datasets_data.h5`.

enjyashraf18 committed:
tables description for elements data

enjyashraf18 committed:
filter data and creating tables

enjyashraf18 committed:
migration script for elements_data

enjyashraf18 committed:
data_info migration

enjyashraf18 committed:
replace pd by the CSV module / renaming

enjyashraf18 committed:
fix chopped cols value / pythonic naming

enjyashraf18 committed:
migration script documentation

enjyashraf18 committed:
migration script documentation

enjyashraf18 committed:
fix data_info empty cols issue

enjyashraf18 committed:
Track HDF5 files with Git LFS

enjyashraf18 committed:
add elements_data HDF5 file

enjyashraf18 committed:
migration script modifications

enjyashraf18 committed:
refactoring species (scalar)

enjyashraf18 committed:
finalizing migration of periodic data

enjyashraf18 committed:
np / pytables version mismatch

enjyashraf18 committed:
data converted to atomic units / change value type in scalar

enjyashraf18 committed:
create Slater Definition Class for run function

enjyashraf18 committed:
create datasets data script

enjyashraf18 committed:
hdf5 creator for slater dataset

enjyashraf18 committed:
remove speciesData / update dump method

enjyashraf18 committed:
refactor load / datafile methods

enjyashraf18 committed:
handle wildcard case while loading the element

enjyashraf18 committed:
remove old periodic dependency

gabrielasd commented on atomdb/datasets/slater/h5file_creator.py:
```diff
@@ -0,0 +1,289 @@
+max_norba = 30 # needs to be calculated
+hdf5_file = files("atomdb.data").joinpath("datasets_data.h5")
+
+SLATER_PROPERTY_CONFIGS = [
```
Include the number of basis functions (`nbasis`) in this list

gabrielasd commented on atomdb/datasets/slater/h5file_creator.py:
```diff
@@ -0,0 +1,289 @@
+    value = pt.Float64Col(shape=(max_norba,))
+
+
+def create_properties_tables(hdf5_file, parent_folder, config, value):
```
Document the code - for this and the functions below, add doctrings explaining what the function does and describing its input parameters

gabrielasd commented on atomdb/datasets/slater/h5file_creator.py:
```diff
@@ -0,0 +1,289 @@
+
+
+def create_tot_array(h5file, parent_folder, key, array_data):
+    tot_gradient_array = h5file.create_carray(parent_folder, key, pt.Float64Atom(), shape=(10000,))
```
The value of the `shape` parameter in this line is hardcoded.
Since it corresponds to the number of radial points where density properties get evaluated, import the variable `NPOINTS` from the `run` script.

https://github.com/enjyashraf18/AtomDB/blob/9be8ea95418584503b818b9cc0360e41a2c0f427/atomdb/datasets/slater/run.py#L37

gabrielasd commented on atomdb/datasets/slater/h5file_creator.py:
```diff
@@ -0,0 +1,289 @@
+
+def create_hdf5_file(fields, dataset, elem, charge, mult, nexc):
+    dataset = dataset.lower()
+    shape = 10000 * max_norba
```
Here also replace the hardcoded value for the number of radial points (NPOINTS=10000)

gabrielasd commented on atomdb/species.py:
```diff
@@ -838,47 +780,44 @@ def compile_species(
-    datapath : str, optional
-        Path to the local AtomDB cache, by default DEFAULT_DATAPATH variable value.
-
+def dump(fields, dataset, elem, charge, mult, nexc):
```
For consistency with the signature of the other functions in this module, I suggest modifying the order of the input arguments to:
`dump(elem, charge, mult, nexc, dataset, fields)`

Or, if it is possible to get the parameters `elem`, `charge`, `mult` and `nexc` from the parameter `fields`, modify the function to:
`dump(fields, dataset)`

enjyashraf18 replied:
Yes, it's definitely possible, instead of passing all the parameters again like in the old structure, we can extract them directly from the fields to make it cleaner

gabrielasd commented on atomdb/species.py:
```diff
@@ -46,6 +51,7 @@
     "at_radius": "at_radius",
     "polarizability": "polarizability",
     "dispersion_c6": "dispersion_c6",
+    "dispersion": "dispersion_c6",
```
Why create duplicated versions of the `dispersion_c6` property?

enjyashraf18 replied:
That was just a temporary fix, when extracting `dispersion` from the `fields` passed by the run modules, it wasn’t being recognized as it exists as `dispersion_c6` in `elements_data.h5`. So I mapped it as a quick fix to test other things. It will be handled properly after the modifications.

gabrielasd commented on atomdb/species.py:
```diff
@@ -778,18 +823,23 @@ def compile_species(
     # Ensure directories exist
     makedirs(path.join(datapath, dataset.lower(), "db"), exist_ok=True)
```
I think this line checking for a `db` directory is no longer necessary.
This folder is for the old database structure where the compiled .msg files got stored under `db`. If the structure of the database has changed to a single HDF5 file containing all compiled data, why check and create a `db` folder?

What should be checked is whether under the folder defined by `datapath` there exists an HDF5 file that contains the specified dataset, and if not, add it

enjyashraf18 replied:
Yes, since it was part of the old structure, I left it as is for now while waiting for our discussion about the data paths (and also the remote URLs later), just to make sure we wouldn’t end up needing any part of it.

gabrielasd commented on atomdb/species.py:
```diff
@@ -778,18 +823,23 @@ def compile_species(
 
-    # print all fields
+
+    # fields = asdict(fields)
```
Remove the comment lines L835-L844

gabrielasd commented on atomdb/species.py:
```diff
@@ -38,6 +38,11 @@
 from numbers import Integral
 
 elements_hdf5_file = files("atomdb.data").joinpath("elements_data.h5")
+datasets_hdf5_file = files("atomdb.data").joinpath("datasets_data.h5")
```
The file `datasets_data.h5` should be placed under `atomdb/datasets`  intead of `atomdb/data`

As a note for the final stage of atomdb refactor, when setting the path where the dataset HDF5 is going to be stored, we should support custom paths; for example if the user wants to compile the dataset outside the atomdb package.

gabrielasd commented on atomdb/species.py:
```diff
@@ -781,19 +739,19 @@ def compile_species(
 
     # dump the data to the HDF5 file
-    dump(fields, dataset, elem, charge, mult, nexc)
+    # dump(fields, dataset, elem, charge, mult, nexc)
```
How does the datasets HDF5 file gets created if the dump step in L742 is commented?
This compile_species function should compile and dump the data for a specified dataset, not return an instance of Species.

enjyashraf18 replied:
Returning `species` and printing the fields was just for quick demos during our meetings (mainly for the [test file](https://github.com/enjyashraf18/AtomDB/blob/refacor-species/atomdb/test.py#L5)). But this part will be removed now, since we've moved past the testing of `compile_species` and the focus is back on loading and dumping the data properly.

enjyashraf18 committed:
add nbasis, use NPOINTS, add docstrings

- Included nbasis in fields
- Replaced hardcoded radial points with imported NPOINTS
- Placed datasets_data.h5 under datasets folder
- Added docstrings

enjyashraf18 committed:
update CI and packaging configs

enjyashraf18 committed:
Improving hdf5 file querying

msricher commented on atomdb/datasets/slater/run.py:
```diff
@@ -39,6 +40,65 @@
 # DATAPATH = os.path.abspath(DATAPATH._paths[0])
 
 
+@dataclass
```
You shouldn't have to use dataclass here, you should be using this instead: https://www.pytables.org/usersguide/libref/declarative_classes.html#isdescriptionclassdescr

(I think, at least, we can cover this tomorrow.)

msricher reviewed (commented):
I think the main functionality here should be good. If you can walk me through it tomorrow, we can fix it up. Thanks for keeping it up while I was away!

enjyashraf18 committed:
add gbasis

enjyashraf18 committed:
converting Slater dataset files from msgpack to HDF5 format
